compiler/pipes/check-classes.cpp

```diff
@@ -79,30 +79,40 @@ void CheckClassesPass::check_serialized_fields(ClassPtr klass) {
-            throw std::invalid_argument("");
-          }
-          kphp_error_return(0 <= kphp_serialized_field && kphp_serialized_field < max_serialization_tag_value, fmt_format("kphp-serialized-field({}) must be >=0 and < than {}", kphp_serialized_field, max_serialization_tag_value + 0));
-          kphp_error_return(!used_serialization_tags_for_fields[kphp_serialized_field], fmt_format("kphp-serialized-field: {} is already in use", kphp_serialized_field));
-          f.serialization_tag = kphp_serialized_field;
-          used_serialization_tags_for_fields[kphp_serialized_field] = true;
-        } catch (std::logic_error &) {
-          kphp_error_return(false, fmt_format("bad kphp-serialized-field: '{}'", *kphp_serialized_field_str));
-        }
+    auto kphp_serialized_field_str = phpdoc_find_tag_as_string(f.phpdoc_str, php_doc_tag::kphp_serialized_field);
+    if (!kphp_serialized_field_str) {
+      kphp_error(!klass->is_serializable, fmt_format("kphp-serialized-field is required for field: {}", f.local_name()));
+      return;
+    }
+
+    kphp_error_return(klass->is_serializable, fmt_format("you may not use @kphp-serialized-field inside non-serializable klass: {}", klass->name));
+    if (vk::string_view(*kphp_serialized_field_str).starts_with("none")) {
+        return;
+    }
+
+    try {
+      size_t processed_pos{0};
+      auto kphp_serialized_field = std::stoi(*kphp_serialized_field_str, &processed_pos);
+      if (processed_pos != kphp_serialized_field_str->size() && vk::none_of_equal((*kphp_serialized_field_str)[processed_pos], ' ', '/', '#', '\n')) {
+        throw std::invalid_argument("");
+      }
+      if (kphp_serialized_field < 0 || max_serialization_tag_value <= kphp_serialized_field) {
+        kphp_error_return(false, fmt_format("kphp-serialized-field({}) must be >=0 and < than {}", kphp_serialized_field, max_serialization_tag_value + 0));
+      }
+      if (used_serialization_tags_for_fields[kphp_serialized_field]) {
+        kphp_error_return(false, fmt_format("kphp-serialized-field: {} is already in use", kphp_serialized_field));
       }
-    } else {
-      kphp_error_return(!klass->is_serializable, fmt_format("kphp-serialized-field is required for field: {}", f.local_name()));
+      f.serialization_tag = kphp_serialized_field;
+      f.serialize_as_float32 = phpdoc_tag_exists(f.phpdoc_str, php_doc_tag::kphp_serialized_float32);
+      used_serialization_tags_for_fields[kphp_serialized_field] = true;
+    } catch (std::logic_error &) {
+      kphp_error_return(false, fmt_format("bad kphp-serialized-field: '{}'", *kphp_serialized_field_str));
     }
   });
 
   klass->members.for_each([&](ClassMemberStaticField &f) {
-    kphp_error_return(!phpdoc_tag_exists(f.phpdoc_str, php_doc_tag::kphp_serialized_field),
+    kphp_error_return(!phpdoc_tag_exists(f.phpdoc_str, php_doc_tag::kphp_serialized_field) &&
+                      !phpdoc_tag_exists(f.phpdoc_str, php_doc_tag::kphp_serialized_float32),
```

Comment:
Why is && here? @kphp-serialized-field must always exist

Reply:
It's about `ClassMemberStaticField` we check here that static fields don't contain `kphp_serialized_field` and `serialized_float32` annotations.